### ISSUE TITLE
Can't attach to process with unmapped executable image

### ISSUE THREAD

**Issue by Mattiwatti** [x64dbg/x64dbg#1622]
- Debugger version: current git (3fe5b0a)
- OS version: XP through 10, 32 and 64 bit
- Plugins: none
- Description of the issue: it's not possible to attach to a process that has its original executable image unmapped from memory using `NtUnmapViewOfSection`.

[POC sample (32 bit)](https://www.dropbox.com/s/kvrbhieuw6a8bva/TryAttach32.exe?dl=1)
[POC sample (64 bit)](https://www.dropbox.com/s/sz1av6q7s11n4fl/TryAttach64.exe?dl=1)

Reproduction steps:
1. Open the sample exe. It will start notepad.exe suspended using `CreateProcess`.
  Note: I recommend using the sample matching the native OS bitness, as WoW64 seems to add some additional complexity to this bug that makes it hard to follow what's going on (it does reproduce, but I'm not 100% sure it's for the same reasons).
2. Press any key to continue, ignoring the first prompt.
3. On the second prompt, try attaching to notepad.exe. This fails with the following log:
```
Database file: C:\Dev\x64dbg\bin\x64\db\notepad.exe.dd64
Attach to process failed! GetLastError() = 998 (ERROR_NOACCESS)
Saving database to C:\Dev\x64dbg\bin\x64\db\notepad.exe.dd64 16ms
Debugging stopped!
```

Notes:
The sample is useless and will always cause notepad to crash with an access violation after its thread is resumed, the obvious result of its image being unmapped. I reduced this POC from a different program to narrow down the behaviour. If you attach on the first prompt (before the image is unmapped), the attach works. You can then see the AV occur in x64dbg after the sample resumes the thread.

WinDbg also has trouble with this, but can recover:
```
*** wait with pending attach

************* Symbol Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*C:\Symbols*https://msdl.microsoft.com/download/symbols
Symbol search path is: srv*C:\Symbols*https://msdl.microsoft.com/download/symbols
Executable search path is: 
ModLoad: 00000000`fffd0000 00000000`fffd0000   notepad.exe
ModLoad: 00000000`76e40000 00000000`76fea000   ntdll.dll
Break-in sent, waiting 30 seconds...
```
(sample resumes notepad.exe)
```
(189c.22e0): Access violation - code c0000005 (first chance)
```

**Comment by blaquee** [comment]
Looks like what you're POC'ing is something malware does very often called Process Hollowing, but usually after NtUnMapViewOfSection is called, the hollowed memory is replaced with some malicious binary code, then it continues execution. (which explains the violation when you try to continue notepad). I've had trouble with this technique with x64dbg as well, but I dont try to attach after the process has been unmapped, Im already usually in a debugging session when it occurs. It is interesting that Windbg recovers, I wonder if its using some kind of file on disk loading.

**Comment by blaquee** [comment]
Oh nevermind, I think I see what WinDBG is doing. When the break-in fails, it resumes the process (thread) then it tries again.

> WARNING: Break-in timed out, suspending.
         This is usually caused by another thread holding the loader lock

That message makes it evident.

**Comment by mrexodia** [comment]
You should be able to use Drew's plugin for child process debugging. That
way you can debug any spawned process before it's loaded.

**Comment by Mattiwatti** [comment]
@blaquee Yep, you're right, process hollowing is exactly what the original executable did/does. Although it was written by myself with no malicious intent (no more than the usual anyway.) That said I did manage to get a 100/100 'malicious' score with it on hybrid-analysis. Ah, anti-virus companies, will they ever learn...

The problem my executable has is the following: if I unmap the original image from the child process, things tend to go poorly (though this seems to be somewhat random - it does occasionally work), even if the original image base and the new one are in no way near each other address-wise. On the other hand, attaching a debugger to a suspended process none of whose threads have ran yet wil have side effects due to the `DbgUiRemoteBreakin` thread ntdll starts (due to it now being the first thread to run in the process). I'm not sure if this problem can really be solved, though WinDbg does have a good recovery considering the messed up state of the process. Maybe I can do something with `WriteProcessMemory` to set ghetto breakpoints without use of a debugger.

@mrexodia Do you mean [DbgChild](https://github.com/David-Reguera-Garcia-Dreg/DbgChild)? Or are Dreg and Drew two different similarly named people? (as that's the only child process debug plugin I could find.) I took a look at it, and **wow** that is a complex plugin o.O I hope nobody tells him about `NtCreateProcess` and `NtCreateProcessEx` for his sake or it will be a full fledged OS soon :P

**Comment by mrexodia** [comment]
@David-Reguera-Garcia-Dreg maybe you can take a look at this? I tried the plugin, but it crashed in the hook or it failed to attach...

**Comment by Mattiwatti** [comment]
Small update on this: it is in fact possible to get the process back into an attachable state after unmapping the original executable. However this does require relocating the replacement image to the base address the original image was loaded at. After adding code to handle the relocation, the 'hollow' executable can now host MATLAB and Photoshop, and x64dbg can attach with no issue. So the key seems to be that any replacement code must have the same base address as the SEC_IMAGE the kernel created the original process with. `SetThreadContext` doesn't help against this, so manual relocation is needed.
